File: 0031-next-permutation/solution.java
class Solution {
    public void nextPermutation(int[] nums) {
        int i;
        for( i=nums.length-2;i>=0;i--){
            if(nums[i]<nums[i+1])
                break;
        }
        if(i<0){
            i=0;
              for(i=0;i<nums.length/2;i++){
            swap(nums,i,nums.length-i-1);
        }
        }
        else{
            int k=i;
            int j;
            for(j=nums.length-1;j>i;j--){
                if(nums[j]>nums[i]){
                    break;
                }
            }
              swap(nums,j,i);
              for(i=k+1;i<nums.length-1;i++){
                  for(int m=i+1;m<nums.length;m++){
                      if(nums[i]>nums[m])
                        swap(nums,i,m);
                
                  }
               }
        }
      
    }
    void swap(int nums[], int j, int i){
        int temp=nums[j];
        nums[j]=nums[i];
        nums[i]=temp;
    }
}

File: 0031-next-permutation/solution.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        //
        int n = nums.size();
        int ind = -1;
        for(int i=n-1; i>0; i--){
            if(nums[i-1] < nums[i]){
                ind = i-1;
                break;
            }
        }
        if(ind == -1){
           reverse(nums.begin(), nums.end());
           return;
        }

        int k = ind+1;
        for(int i=k+1; i<n; i++){
            if(nums[i] > nums[ind] && nums[i] < nums[k]) k = i;
        }

        swap(nums[ind], nums[k]);
        sort(nums.begin()+ind+1, nums.end());

        return;
    }
};
